cores/arduino/wiring.c

```diff
@@ -22,44 +18,126 @@
 
 #include "wiring_private.h"
 
-// the prescaler is set so that timer0 ticks every 64 clock cycles, and the
+// CHANGE by Gijs Withagen, 18-02-2019
+// Prescaler can be changed in runtime and still have a correct millis()
+// the micros() just has a glitch, but counted form changing the prescaler
+// it gives proper relative values.
+//
+// Defalult the prescaler is set so that timer0 ticks every 64 clock cycles, and the
 // the overflow handler is called every 256 ticks.
-#define MICROSECONDS_PER_TIMER0_OVERFLOW (clockCyclesToMicroseconds(64 * 256))
+#define MICROSECONDS_PER_TIMER0_OVERFLOW (clockCyclesToMicroseconds(getPrescaler() * 256))
 
 // the whole number of milliseconds per timer0 overflow
 #define MILLIS_INC (MICROSECONDS_PER_TIMER0_OVERFLOW / 1000)
 
-// the fractional number of milliseconds per timer0 overflow. we shift right
-// by three to fit these numbers into a byte. (for the clock speeds we care
-// about - 8 and 16 MHz - this doesn't lose precision.)
-#define FRACT_INC ((MICROSECONDS_PER_TIMER0_OVERFLOW % 1000) >> 3)
-#define FRACT_MAX (1000 >> 3)
+// the fractional number of milliseconds per timer0 overflow.
+#define FRACT_INC ((MICROSECONDS_PER_TIMER0_OVERFLOW % 1000))
+#define FRACT_MAX (1000)
 
 volatile unsigned long timer0_overflow_count = 0;
-volatile unsigned long timer0_millis = 0;
-static unsigned char timer0_fract = 0;
+
+volatile unsigned long correction_millis = 0;
+volatile unsigned long correction_frac = 0;
+
+volatile unsigned long millis_correction_inc = 0;
+volatile unsigned long fract_correction_inc = 0;
+volatile unsigned int previous_prescaler = 0;
+
+// Returns the actual value of the prescaler of Timer 0 needed to calculate
+// the elapsed microseconds and milliseconds based on the timer0_overflow_count
+unsigned int getPrescaler() {
+	// set timer 0 prescale factor to 64
+#if defined(__AVR_ATmega128__)
+	// CPU specific: different values for the ATmega128
+	switch ((TTCR0 & _BV(CS02)) + (TTCR0 & _BV(CS01)) + (TCCR0 & _BV(CS00))) {
```

Comment:
To me, it reads more easily when I can calculate the mask and apply it in one go:

```
#if defined(__AVR_ATmega128__)
#        define PRESCALER0_MASK (_BV(CS02) | _BV(CS01) | _BV(CS00))
#        define PRESCALER0_REG TTCR0
#elif defined (__AVR_ATmega8__)
#        define PRESCALER0_MASK (_BV(CS02) | _BV(CS01) | _BV(CS00))
#        define PRESCALER0_REG TTCR0
#elif defined(__AVR_ATmega168__) \
    || defined(__AVR_ATmega328) || ...
#        define PRESCALER0_MASK (_BV(CS02) | _BV(CS01) | _BV(CS00))
#        define PRESCALER0_REG TCCR0B
#elif defined(__AVR_ATmega645__)
#        define PRESCALER0_MASK (_BV(CS02) | _BV(CS01) | _BV(CS00))
#        define PRESCALER0_REG TTCRA
#else
#  error Timer 0 prescaler cannot be determined correctly
#endif
```

Then, the prescaler bit pattern can be defined generally, for all architectures.

```
#define PRESCALER0_BITS (PRESCALER0_REG & PRESCALER0_MASK)
```

The value seems to be platform dependent, too, and since all compilers I know of understand binary literal, the real bit pattern can be expressed more directly:

```
....
    switch(PRESCALER0_BITS) {
#if defined(__AVR_ATmega128__)
      case 0b000: return 1;
      case 0b001: return 1;
      case 0b010: return 8;
      case 0b011: return 32;
      case 0b100: return 64;
      case 0b101: return 128;
      case 0b110: return 256;
      case 0b111: return 1024;
 #else
      case 0b000: return 1;
      case 0b001: return 1;
      case 0b010: return 8;
      case 0b011: return 64;
      case 0b100: return 256;
      case 0b101: return 1024;
      case 0b110: return 1;
      case 0b111: return 1;
#endif
    }
```